File: src/main/java/io/tolgee/controllers/RepositoryController.java
package io.tolgee.controllers;

import io.tolgee.dtos.request.CreateRepositoryDTO;
import io.tolgee.dtos.request.EditRepositoryDTO;
import io.tolgee.dtos.request.InviteUser;
import io.tolgee.dtos.response.RepositoryDTO;
import io.tolgee.exceptions.InvalidStateException;
import io.tolgee.exceptions.NotFoundException;
import io.tolgee.model.Permission;
import io.tolgee.model.Repository;
import io.tolgee.model.UserAccount;
import io.tolgee.security.AuthenticationFacade;
import io.tolgee.service.InvitationService;
import io.tolgee.service.RepositoryService;
import io.tolgee.service.SecurityService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.*;

import javax.validation.Valid;
import java.util.Set;

@RestController("_repositoryController")
@CrossOrigin(origins = "*")
@RequestMapping("/api/repositories")
public class RepositoryController implements IController {

    private final RepositoryService repositoryService;
    private final AuthenticationFacade authenticationFacade;
    private final SecurityService securityService;
    private final InvitationService invitationService;

    @Autowired
    public RepositoryController(RepositoryService repositoryService, AuthenticationFacade authenticationFacade, SecurityService securityService, InvitationService invitationService) {
        this.repositoryService = repositoryService;
        this.authenticationFacade = authenticationFacade;
        this.securityService = securityService;
        this.invitationService = invitationService;
    }

    @PostMapping(value = "")
    public RepositoryDTO createRepository(@RequestBody @Valid CreateRepositoryDTO dto) {
        UserAccount userAccount = authenticationFacade.getUserAccount();
        Repository repository = repositoryService.createRepository(dto, userAccount);
        return RepositoryDTO.fromEntityAndPermission(repository, repository.getPermissions().stream().findAny().orElseThrow(InvalidStateException::new));
    }

    @GetMapping(value = "/{id}")
    public RepositoryDTO getRepository(@PathVariable("id") Long id) {
        Permission permission = securityService.getAnyRepositoryPermission(id);
        return RepositoryDTO.fromEntityAndPermission(repositoryService.getById(id).orElseThrow(null), permission);
    }


    @PostMapping(value = "/edit")
    public RepositoryDTO editRepository(@RequestBody @Valid EditRepositoryDTO dto) {
        Permission permission = securityService.checkRepositoryPermission(dto.getRepositoryId(), Permission.RepositoryPermissionType.MANAGE);
        Repository repository = repositoryService.editRepository(dto);
        return RepositoryDTO.fromEntityAndPermission(repository, permission);
    }

    @GetMapping(value = "")
    public Set<RepositoryDTO> getAll() {
        return repositoryService.findAllPermitted(authenticationFacade.getUserAccount());
    }

    @DeleteMapping(value = "/{id}")
    public void deleteRepository(@PathVariable Long id) {
        securityService.checkRepositoryPermission(id, Permission.RepositoryPermissionType.MANAGE);
        repositoryService.deleteRepository(id);
    }

    @PostMapping("/invite")
    public String inviteUser(@RequestBody InviteUser invitation) {
        securityService.checkRepositoryPermission(invitation.getRepositoryId(), Permission.RepositoryPermissionType.MANAGE);
        Repository repository = repositoryService.getById(invitation.getRepositoryId()).orElseThrow(NotFoundException::new);
        return invitationService.create(repository, invitation.getType());
    }
}

File: src/main/java/io/tolgee/controllers/RepositoryController.kt
package io.tolgee.controllers

import io.swagger.v3.oas.annotations.Operation
import io.swagger.v3.oas.annotations.tags.Tag
import io.tolgee.configuration.tolgee.TolgeeProperties
import io.tolgee.dtos.request.CreateProjectDTO
import io.tolgee.dtos.request.EditProjectDTO
import io.tolgee.dtos.request.ProjectInviteUserDto
import io.tolgee.dtos.response.ProjectDTO
import io.tolgee.dtos.response.ProjectDTO.Companion.fromEntityAndPermission
import io.tolgee.exceptions.NotFoundException
import io.tolgee.exceptions.PermissionException
import io.tolgee.model.Permission
import io.tolgee.model.UserAccount
import io.tolgee.security.AuthenticationFacade
import io.tolgee.service.InvitationService
import io.tolgee.service.PermissionService
import io.tolgee.service.ProjectService
import io.tolgee.service.SecurityService
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.web.bind.annotation.*
import javax.validation.Valid

@RestController("_projectController")
@CrossOrigin(origins = ["*"])
@RequestMapping("/api/projects")
@Tag(name = "Project")
open class ProjectController @Autowired constructor(private val projectService: ProjectService,
                                                       private val authenticationFacade: AuthenticationFacade,
                                                       private val securityService: SecurityService,
                                                       private val invitationService: InvitationService,
                                                       private val permissionService: PermissionService,
                                                       private val tolgeeProperties: TolgeeProperties
) : IController {

    @PostMapping(value = [""])
    @Operation(summary = "Creates project with specified languages")
    open fun createProject(@RequestBody @Valid dto: CreateProjectDTO?): ProjectDTO {
        val userAccount = authenticationFacade.userAccount
        if (!this.tolgeeProperties.authentication.userCanCreateProjects
                && userAccount.role != UserAccount.Role.ADMIN) {
            throw PermissionException()
        }
        val project = projectService.createProject(dto!!)
        val type = permissionService.getProjectPermissionType(project.id, userAccount)
                ?: throw IllegalStateException()
        return fromEntityAndPermission(project, type)
    }

    @GetMapping(value = ["/{id}"])
    @Operation(summary = "Returns project by id")
    open fun getProject(@PathVariable("id") id: Long?): ProjectDTO {
        val permission = securityService.checkAnyProjectPermission(id!!)
        val project = projectService.get(id).orElseThrow { NotFoundException() }!!
        return fromEntityAndPermission(project, permission)
    }

    @Operation(summary = "Modifies project")
    @PostMapping(value = ["/edit"])
    open fun editProject(@RequestBody @Valid dto: EditProjectDTO?): ProjectDTO {
        val permission = securityService.checkProjectPermission(dto!!.projectId!!, Permission.ProjectPermissionType.MANAGE)
        val project = projectService.editProject(dto)
        return fromEntityAndPermission(project, permission)
    }

    @GetMapping(value = [""])
    @Operation(summary = "Return all projects, where use has any access")
    open fun getAll(): List<ProjectDTO> = projectService.findAllPermitted(authenticationFacade.userAccount)

    @DeleteMapping(value = ["/{id}"])
    @Operation(summary = "Deletes project by id")
    open fun deleteProject(@PathVariable id: Long?) {
        securityService.checkProjectPermission(id!!, Permission.ProjectPermissionType.MANAGE)
        projectService.deleteProject(id)
    }

    @PostMapping("/invite")
    @Operation(summary = "Generates user invitation link for project")
    open fun inviteUser(@RequestBody @Valid invitation: ProjectInviteUserDto): String {
        securityService.checkProjectPermission(invitation.projectId!!, Permission.ProjectPermissionType.MANAGE)
        val project = projectService.get(invitation.projectId!!).orElseThrow { NotFoundException() }!!
        return invitationService.create(project, invitation.type!!)
    }
}
